idl/internal/thrift.y

```diff
@@ -268,38 +270,65 @@ enum_item
     ;
 
 fields
-    : /* nothing */ { $$ = nil }
+    : /* nothing */             { $$ = nil }
     | fields field optional_sep { $$ = append($1, $2) }
     ;
 
-
 field
-    : lineno docstring INTCONSTANT ':' field_required type IDENTIFIER type_annotations
+    : lineno docstring field_identifier field_required type IDENTIFIER type_annotations
         {
             $$ = &ast.Field{
-                ID: int($3),
-                Name: $7,
-                Type: $6,
-                Requiredness: $5,
-                Annotations: $8,
+                ID: $3,
+                Name: $6,
+                Type: $5,
+                Requiredness: $4,
+                Annotations: $7,
                 Line: $1,
                 Doc: ParseDocstring($2),
             }
         }
-    | lineno docstring INTCONSTANT ':' field_required type IDENTIFIER '=' const_value
-      type_annotations
+    | lineno docstring field_required type IDENTIFIER type_annotations
         {
             $$ = &ast.Field{
-                ID: int($3),
-                Name: $7,
-                Type: $6,
-                Requiredness: $5,
-                Default: $9,
-                Annotations: $10,
+                IDUnset: true,
+                Name: $5,
+                Type: $4,
+                Requiredness: $3,
+                Annotations: $6,
                 Line: $1,
                 Doc: ParseDocstring($2),
             }
         }
+    | lineno docstring field_identifier field_required type IDENTIFIER '=' const_value type_annotations
+        {
+            $$ = &ast.Field{
+                ID: $3,
+                Name: $6,
+                Type: $5,
+                Requiredness: $4,
+                Default: $8,
+                Annotations: $9,
+                Line: $1,
+                Doc: ParseDocstring($2),
+            }
+        }
+    | lineno docstring field_required type IDENTIFIER '=' const_value type_annotations
+        {
+            $$ = &ast.Field{
+                IDUnset: true,
+                Name: $5,
+                Type: $4,
+                Requiredness: $3,
+                Default: $7,
+                Annotations: $8,
+                Line: $1,
+                Doc: ParseDocstring($2),
+            }
+        }
+    ;
+
+field_identifier
+    : INTCONSTANT ':' { $$ = int($1) }
```

Comment:
I'm not sure if we can do this, but would it makes sense to have this handle the empty case and return both, the field ID and whether it's unset? Something like this:

```
type fieldIdentifier struct {
  ID int
  Unset bool
}
```

And then,

```suggestion
field_identifier
    : INTCONSTANT ':' { $$ = fieldIdentifier{ID: int($1)} }
    | /* na */        { $$ = fieldIdentifier{Unset: true} }
```

That might let us half the number of cases above as well.

```
$$ = &ast.Field{
  // ...,
  ID: $3.ID,
  IDUnset: $3.Unset,
}
```

Reply:
I didn't realize this syntax would be so straightforward. Thanks!